### PR TITLE
Blueprint/train car operations UI window

### PR DESCRIPTION
**Pull request by mbm-OR**
Blueprint https://blueprints.launchpad.net/or/+spec/traincaroperations-ui

More info: http://www.elvastower.com/forums/index.php?/topic/37062-proposal-f9-train-operations/page__view__findpost__p__301151

### CODE REVIEW

**Comment by mbm-OR** [comment]
Hi Siebren,
You have reacted very quickly. I have tested your commit and it works fine. Thanks for your cooperation.

**Comment by sweiland-openrails** [comment]
hi Mauricio,

I see a few problems and have one suggestion:

- first click on car does not open sub window, the second click does (or is this on purpose?)
- closing with ctrl-alt-f9 does not close sub window
- when clicking on left coupler wrong car is uncoupled, makes it also impossible to uncouple last car
- first click on a car gives the wrong camera view, always the first car
- I would only change the camera view when Camera in Outside Front or Rear view. Not when in Cab view. But that's my personal opinion.

regards, Siebren.

**Comment by mbm-OR** [comment]
Hi Siebren,
- The first click sets the window size and the last visible carID for vertical scrolling. It is needed at this moment.
- To close the "Train Car Viewer" pop-up, click on the carID.
- The front coupler bug has been fixed. Thanks for finding it.
- The first click on a carID gives the front car of the selected car.
- The purpose of this improvement is to focus on train operations where the external camera seems to be needed very often. If users consider to keep the cab or 3D-cab views, the current procedure can be changed without any problems.

Regards,
Mauricio

**Comment by sweiland-openrails** [comment]
hi Mauricio,

thanks for the fix. And I agree with the rest of your comments. It's a really nice and large piece of code you've written for this enhancement.

I'm looking into adding the web interface for this enhancement.

regards, Siebren.

**Comment by mbm-OR** [comment]
The current fix allows:
- When uncoupling a car, the new last car is selected by default.
- Fixed a bug with the vertical scroll window.
- Unused code removed.

The camera is not pointing correctly at the selected car, if there are cars of different lengths in the consist. This bug was also reported by Siebren. I'll have to study it.

**Comment by mbm-OR** [comment]
Hi Siebren,
Your contribution to the manual is very clear and concise.
Thank you for your work.

File: Source/RunActivity/Viewer3D/Popups/TrainCarOperationsWindow.cs [hunk 1]
```diff
@@ -0,0 +1,1112 @@
...
+                            }
+                            else
+                            {
+                                line.Add(new buttonFrontBrakeHose(0, 0, SymbolSize, Owner.Viewer, car, carPosition));
+                                line.Add(new buttonFrontAngleCock(0, 0, SymbolSize, Owner.Viewer, car, carPosition));
+                                AddSpace();
+                                line.Add(new buttonCouplerFront(0, 0, SymbolSize, Owner.Viewer, car));
+                            }
+                            line.Add(carLabel);
+                            if (car != PlayerTrain.Cars.Last())
```
Comment (sweiland-openrails) [inline Source/RunActivity/Viewer3D/Popups/TrainCarOperationsWindow.cs:398]: looks to me the if and the else clause contain the same statements

**Comment by mbm-OR** [comment]
Hi Siebren,
After your[ improvement](https://github.com/openrails/openrails/pull/882/commits/9df6e1ddc058785b810c6cda0c0dd8a9b07e4f85), the web-page scrolls quickly to synchronise with the main window. It looks great.

**Comment by sweiland-openrails** [comment]
hi Mauricio,

I hope I've solved the merge conflicts properly. I'll test the unstable once it's available.
And in the first commit I replaced some code of yours, hope you don't mind.
I have no more enhancements/bugs for this topic.

Are you planning to swap the F9 and Ctrl-Alt-F9 buttons? So that the F9 is for the enhanced Train Car Operations window. And the Ctrl-Alt-F9 for the users who like the old format. 

regards, Siebren.

**Comment by mbm-OR** [comment]
Hi Siebren,
When using this new Train Car Operations feature, it was sometimes difficult to see the car ID labels.
To make it easier, the car ID label is now visible on top of the selected car. 
This improvement only works, if we have switched to Player Control, the Car labels <F7> are visible and the Train Operations Viewer or the Train Car Operations Webpage is visible, otherwise it works as usual. 
Any feedback is appreciated.

**Comment by mbm-OR** [comment]
Added fix for [this](https://www.elvastower.com/forums/index.php?/topic/37765-or-forgets-target-car-u20240124-2007/page__view__findpost__p__304984) issue.

**Comment by mbm-OR** [comment]
In the next fix, pressing Alt+1, Alt+3, Alt+7 or Alt+9 (camera keys) will also scroll the layout.

**Comment by mbm-OR** [comment]
[Avoids](https://github.com/openrails/openrails/pull/882/commits/3ca0eb19f513a1331c05cd6474bb0a31bec18b6c) crash when the PlayerTrain was changed from the Train List window.

**Comment by SteelFill** [comment]
While testing the PR to improve the built in graphics #924 I couldn't help but notice the sheer number of .png files this PR has added to the Content folder. For the sake of reducing clutter, would you be willing to move the TrainOperations*.png files to a dedicated folder, perhaps "Content\TrainOperations"?

There are probably other ways to reduce the number of textures, like putting these on an atlas instead of having them all separate but organizing them into a folder would be a good place to start.

**Comment by mbm-OR** [comment]
> ... would you be willing to move the TrainOperations*.png files to a dedicated folder, perhaps "Content\TrainOperations"?
> ... organising them into a folder would be a good place to start.

This suggestion makes sense. I'll take a look.

**Comment by mbm-OR** [comment]
> There are probably other ways to reduce the number of textures, like putting these on an atlas instead of having them all separate ...

This was implemented since the build U2024.05.16-1422.

**Comment by SteelFill** [comment]
> > There are probably other ways to reduce the number of textures, like putting these on an atlas instead of having them all separate ...
> 
> This was implemented since the build U2024.05.16-1422.

Nicely done. All tidy now!

Are you considering changing the keybind of this menu from the cumbersome Ctrl + Alt + F9 combination to just F9 (ie: replacing the old default)? It would be a bit weird to go through the whole process of making a replacement just to leave the old one in charge.

**Comment by mbm-OR** [comment]
> Nicely done. All tidy now!

Thanks.

> Are you considering changing the keybind of this menu from the cumbersome Ctrl + Alt + F9 combination to just F9 (ie: replacing the old default)? It would be a bit weird to go through the whole process of making a replacement just to leave the old one in charge.

I agree with you.
In this way, we swap the F9 key for the enhanced **Train Car Operations** window and the Ctrl+Alt+F9 keys for those users who prefer the old format.
This key assignment change is possible, but must be approved by the ORMT (@cjakeman / @twpol).

**Comment by cjakeman** [comment]
Hi Mauricio,

> > Are you considering changing the keybind of this menu from the cumbersome Ctrl + Alt + F9 combination to just F9 (ie: replacing the old default)? It would be a bit weird to go through the whole process of making a replacement just to leave the old one in charge.
> 
> I agree with you. In this way, we swap the F9 key for the enhanced **Train Car Operations** window and the Ctrl+Alt+F9 keys for those users who prefer the old format. This key assignment change is possible, but must be approved by the ORMT (@cjakeman / @twpol).

Makes sense. I've made a note to raise this at our next session Sat 8th June.

Thanks,

Chris

**Comment by cjakeman** [comment]
Hi Mauricio,

> > Are you planning to swap the F9 and Ctrl-Alt-F9 buttons? So that the F9 is for the enhanced Train Car Operations window. And the Ctrl-Alt-F9 for the users who like the old format.
> 
> This swap would aim to give priority to this improvement over the old format. We should let the **ORMT** (@cjakeman) consider to swap the **F9** key for the enhanced Train Car Operations window and the **Ctrl+Alt+F9** keys, for the users who like the old format. Similar situation has happened in the past with a previous enhancement, such as the Train Driving feature.

I raised this suggestion at our ORMT session as promised.

Please use the F9 key for the new format. That will make it visible to users of the Unstable Version who may not have noticed it yet.

I am assuming that the new format does not remove any functionality provided by the old format. James and I think that this PR should remove the old format entirely. If there are non-trivial objections, we can bring it back, but I think it is better for Open Rails if there is a single way to do things.

Hope that helps and I'm sorry this has taken so long to get to ORMT,

Chris

**Comment by SteelFill** [comment]
I still find it unintuitive that the train operations viewer window needs to be closed by clicking on the part of the window with the car ID. Could it be changed so that closing the train car operations window with F9 also closes any viewer window? (Right now, hitting F9 with a train operations viewer window open just leaves the viewer window there.)

**Comment by mbm-OR** [comment]
> Could it be changed so that closing the train car operations window with F9 also closes any viewer window?

It makes sense. We need to make everything as intuitive as possible.

File: Source/RunActivity/Viewer3D/Viewer.cs [hunk 2]
```diff
@@ -992,6 +1012,22 @@ void HandleUserInput(ElapsedTime elapsedTime)
             }
             if (UserInput.IsPressed(UserCommand.DisplaySwitchWindow)) if (UserInput.IsDown(UserCommand.DisplayNextWindowTab)) SwitchWindow.TabAction(); else SwitchWindow.Visible = !SwitchWindow.Visible;
             if (UserInput.IsPressed(UserCommand.DisplayTrainOperationsWindow)) if (UserInput.IsDown(UserCommand.DisplayNextWindowTab)) TrainOperationsWindow.TabAction(); else { TrainOperationsWindow.Visible = !TrainOperationsWindow.Visible; if (!TrainOperationsWindow.Visible) CarOperationsWindow.Visible = false; }
+
+            if (UserInput.IsPressed(UserCommand.DisplayTrainCarOperationsWindow))
+            {
+                if (UserInput.IsDown(UserCommand.DisplayNextWindowTab))
+                    TrainCarOperationsWindow.TabAction();
+                else
```
Comment (sweiland-openrails) [inline Source/RunActivity/Viewer3D/Viewer.cs:1020]: hi Mauricio (@mbm-OR),

I think this `TrainCarOperationsWindow.TabAction(); ` is of no use. First of all there is no TabAction for TrainCarOperationsWindow. Secondly the UserCommand.DisplayNextWindowTab is only True when the Shift button is pressed. But Shift+F9 fires the "Display Train Dpu Window". So this TabAction code never gets fired.

Shall I remove it? 

I have more or less the same problem with UserCommand.DisplayTrainDpuWindow. Which does not pop up when activated from my web based Switch Panel. Trying to fix it gave a merge conflict with this code. I would like to fix this also in this PR.

Anyway complicated, I hope you understand.

Comment (mbm-OR) [reply Source/RunActivity/Viewer3D/Viewer.cs]: Hi Sierben, ( @sweiland-openrails ) 

> Shall I remove it?
> 
> I have more or less the same problem with UserCommand.DisplayTrainDpuWindow. Which does not pop up when activated from my web based Switch Panel.

All unused TabAction() have been removed. 
Your web based Switch Panel now seems to work with the UserCommand.DisplayTrainDpuWindow.

**Review by sweiland-openrails — APPROVED** [review]
Thanks for the changes, is exactly what I meant

**Comment by sweiland-openrails** [comment]
hi @mbm-OR, since you swapped the buttons to start this enhanced train car operations, could you merge this fix? Looks like people do want to keep the old form of this window next to enhanced one. Or do you have to ask Chris Jakeman first?

I'm asking for this merge also because I want to create some enhancements for the switch panel . However I'm getting a merge conflict in unstable due to this PR.

regards, Siebren.

**Comment by cjakeman** [comment]
Hi Mauricio,

Is this PR complete now?

I think you will have to mark the "Unresolved conversations" as resolved as these are blocking the process for review. Once reviewed, we can get it approved and merged.

Thanks,

Chris

**Comment by mbm-OR** [comment]
> Is this PR complete now?
>
Depending on the user's preference, both the old and new versions of "**Train Car Operations**" are still available in this PR.

> I think you will have to mark the "Unresolved conversations" as resolved as these are blocking the process for review. Once reviewed, we can get it approved and merged.
> 

Thank you for pointing out the "Unresolved  conversations".

Regards,
Mauricio

**Pull request merged**